### DISCUSSION TITLE
Union between Valid type should be Valid

### DISCUSSION THREAD

**Discussion by ihakh** [fastapi/fastapi#8963] [Questions] [answered]
### First check

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Example

```Python
from fastapi import FastAPI, Query
from enum import Enum
from typing import List, Union

app = FastAPI()


class Tag(str, Enum):
    a = 'a'
    b = 'b'
    c = 'c'


@app.get("/")
async def read_file(tag_or_list: Union[Tag, List[str]] = Query(...)):
    return tag_or_list

```

### Description
I got the following error for running `uvicorn test:app`:
```
  File "/usr/local/lib/python3.8/site-packages/fastapi/dependencies/utils.py", line 333, in get_dependant
    assert isinstance(
AssertionError: Param: tag_or_list can only be a request body, using Body(...)
```
a `List[str]` is valid and `Enum` instace for `Query` is valid why `Union` between theme is not valid?

### Environment

* OS: Linux
* FastAPI Version : 0.61.2
* Python version: 3.8.6

**Comment by ycd** [comment, marked as answer]
In the example, `tag_or_list` is just a **one** query parameter, but since you are giving **more than one** Models to it with `Union`, it understands the situation and assumes it is a **Body** that's why it expects a `Body`

You can send an array as a query parameter like this `/products?color=blue,green,red`  but you have one query parameter in this case. So you can not serialize two different things in one query parameter in any way.

So there are two options for you.

You can create a new model.

```python
from fastapi import FastAPI, Depends
from enum import Enum
from pydantic import BaseModel
from typing import List

app = FastAPI()


class Tag(str, Enum):
    a = "a"
    b = "b"
    c = "c"


class NewModel(BaseModel):
    tag: Tag
    my_list: List[str]


@app.post("/")
async def read_file(tag_or_list: NewModel = Depends()):
    return tag_or_list
```

Or you can create two different query parameters.

```python
from fastapi import FastAPI, Query
from enum import Enum
from pydantic import BaseModel
from typing import List

app = FastAPI()


class Tag(str, Enum):
    a = "a"
    b = "b"
    c = "c"


@app.post("/")
async def read_file(tag: Tag, my_list: List[str] = Query(...)):
    return tag_or_list
```

There is no way to merge two different things in one Query parameter, if you need one result, you need to use a Request Body.

**Comment by ycd** [comment]
Is your issue is resolved @ihakh ?

**Comment by david-shiko** [comment]
I had the same problem and I'm wondering what is a problem to check if you got only `tag` or only `list` but not two or multiple together.
I'm not expecting validation or some special behavior, but just to get generate auto docs which will say "Please don't use these parameters together, anyway will work only one of them".
Because of this reason, I'm considering that a question should be opened.
